File: desktop/src/main/java/com/middin/innovatie/desktop/LocalDevAccounts.java
package com.middin.innovatie.desktop;

import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/** Offline dev accounts from resources/dev-accounts.json (source: dev-accounts/dev-accounts.json). */
public final class LocalDevAccounts {
    private static final String RESOURCE_PATH = "/dev-accounts.json";
    private static final Pattern ACCOUNT_BLOCK = Pattern.compile(
            "\\{\\s*\"username\"\\s*:\\s*\"((?:\\\\.|[^\"\\\\])*)\"\\s*,\\s*\"password\"\\s*:\\s*\"((?:\\\\.|[^\"\\\\])*)\"\\s*\\}");

    private static volatile List<Account> cached;

    private LocalDevAccounts() {}

    public static boolean matches(String username, String password) {
        if (username == null || password == null) return false;
        String u = username.trim();
        if (u.isEmpty() || password.isEmpty()) return false;
        for (Account account : all()) {
            if (account.username.equalsIgnoreCase(u) && account.password.equals(password)) {
                return true;
            }
        }
        return false;
    }

    public static List<Account> all() {
        List<Account> list = cached;
        if (list != null) return list;
        synchronized (LocalDevAccounts.class) {
            if (cached != null) return cached;
            cached = Collections.unmodifiableList(loadAccounts());
            return cached;
        }
    }

    public static List<String> usernames() {
        List<String> names = new ArrayList<>();
        for (Account account : all()) {
            names.add(account.username);
        }
        return names;
    }

    private static List<Account> loadAccounts() {
        try (InputStream in = LocalDevAccounts.class.getResourceAsStream(RESOURCE_PATH)) {
            if (in == null) return List.of();
            String json = new String(in.readAllBytes(), StandardCharsets.UTF_8);
            if (!json.isEmpty() && json.charAt(0) == '\uFEFF') {
                json = json.substring(1);
            }
            return parseAccounts(json);
        } catch (IOException e) {
            return List.of();
        }
    }

    static List<Account> parseAccounts(String json) {
        List<Account> out = new ArrayList<>();
        Matcher m = ACCOUNT_BLOCK.matcher(json);
        while (m.find()) {
            String username = unescapeJson(m.group(1)).trim();
            String password = unescapeJson(m.group(2));
            if (!username.isEmpty() && !password.isEmpty()) {
                out.add(new Account(username, password));
            }
        }
        return out;
    }

    private static String unescapeJson(String s) {
        return s.replace("\\\"", "\"").replace("\\\\", "\\");
    }

    public record Account(String username, String password) {
        public Account {
            username = username.toLowerCase(Locale.ROOT);
        }
    }
}

File: app/src/main/java/com/middin/innovatie/app/auth/LocalDevAccounts.kt
package com.middin.innovatie.app.auth

import android.content.Context
import org.json.JSONObject
import java.io.IOException

/** Offline dev accounts from assets/dev-accounts.json (source: dev-accounts/dev-accounts.json). */
object LocalDevAccounts {
  private const val ASSET_PATH = "dev-accounts.json"

  fun matches(context: Context, username: String, password: String): Boolean {
    if (username.isBlank() || password.isBlank()) return false
    val u = username.trim()
    return loadAccounts(context).any { (name, pass) ->
      name.equals(u, ignoreCase = true) && pass == password
    }
  }

  fun usernames(context: Context): List<String> =
    loadAccounts(context).map { it.first }

  private fun loadAccounts(context: Context): List<Pair<String, String>> {
    val json = try {
      context.assets.open(ASSET_PATH).bufferedReader().use { it.readText() }
    } catch (_: IOException) {
      return emptyList()
    }
    return parseAccounts(json)
  }

  internal fun parseAccounts(jsonBody: String): List<Pair<String, String>> {
    val root = JSONObject(jsonBody.trim().removePrefix("\uFEFF"))
    val arr = root.optJSONArray("accounts") ?: return emptyList()
    val out = mutableListOf<Pair<String, String>>()
    for (i in 0 until arr.length()) {
      val item = arr.optJSONObject(i) ?: continue
      val username = item.optString("username", "").trim()
      val password = item.optString("password", "")
      if (username.isNotEmpty() && password.isNotEmpty()) {
        out.add(username to password)
      }
    }
    return out
  }
}
